### DISCUSSION TITLE
x-model.number is too strict with decimal separators

### DISCUSSION THREAD

**Discussion by ghost** [alpinejs/alpine#2184] [7. Bugs (archived)] [answered]
Hy!

Thanks for this awesome library it helped me out with lots of projects.
While I was working on a financial web application I noticed that the ".number" modifier takes the Javascript number typecasting a bit too strictly.

[jsfiddle](https://jsfiddle.net/TKrisee/8xdpn14k/)

I know that for decimals in Javascript the separator has to be a period, but many spoken languages (like Hungarian, Arabic, etc.) use a comma as a decimal separator. 

My main problem with this is that mobile numeric keyboards will show the separator native to the language set for the device itself what might either be a comma or a period. I know that I can drop the ".number" modifier write some custom logic where I replace the comma for a period and cast that to a Javascript number but this seems like a hack for me at least while I'm using alpine what was made for things like these while it also makes the .number unusable for localized sites.

**Comment by SimoTod** [comment, marked as answer]
I think it's more of a browser issue. 
Many browsers don't support commas as a decimal separator when the input has type="number" and js only allows float with periods independently by the language. 
Doesn't your mobile keyboard show both comma and period?
If you want to allow commas, maybe you don't need the numeric modifiers.

**Reply by ghost** [reply]
I try to avoid using the 'input type="number"' as much as possible as it's inconsistent and can be used only for incremental numbers.

"Many browsers don't support commas as a decimal separator when the input has type="number" and js only allows float with periods independently by the language."

While the Javascript part is true what I did mention in the issue, the former is simply just not true(chrome, safari, edge are all allowing both commas and periods as decimal separatos, it's simply based on the localization). And even if it were it is still not an acceptable approach to tell the user to use another browser. I mean if we were talking about IE I'd let it pass but this is not the case.

"Doesn't your mobile keyboard show both comma and period?" - sadly on iOS, it does not.

I want to enable commas because from the user's perspective who's not a native English speaker commas ARE the decimal separators.

**Reply by SimoTod** [reply]
Try to type a comma here: https://www.w3schools.com/tags/tryit.asp?filename=tryhtml5_input_type_number 
I'm on chrome mobile and it doesn't allow me which means that my browser doesn't support it, what you see in the keyboard ia your os, not the browser.

**Reply by SimoTod** [reply]
And, I'm from a country where the decimal separator is a comma, just to clarify

**Reply by ghost** [reply]
> Try to type a comma here: https://www.w3schools.com/tags/tryit.asp?filename=tryhtml5_input_type_number
> I'm on chrome mobile and it doesn't allow me which means that my browser doesn't support it, what you see in the keyboard ia your os, not the browser.

I know that for the input type="number" some mobile browser won't allow a comma thats why I did not use it in my jsfiddle. Because it's inconsistent won't show the correct keyboard type (I mean I get to type in all sorts of special symbols on iOS whereas the input only accepts numbers, exponential numbers, periods and negative signs. Not to mention the horrible accesibility it has.

If you check my fiddle you can see that I used inputmode="decimal" instead what is (in genaral) better to be used for anything that is numeric but not strictly a number.

Still my point is that I think alpine should generaly handle all decimal numbers with the .number modifer correctly with broader limitations for the separator.

**Reply by SimoTod** [reply]
Yeah, my original sentence specified 'using type number' and you said it wasn't true. I understand that you don't use it and i get that it offers a poor user experience, but that's what a browser considers as a number. 

By the way, the problem is that `.number` casts to... Well... a js number and a period is the only allowed decimal separator.

The project is open source so you can try to implement what you described but the property in x-data will end up using a period so when x-model updates, itwill change what the user entered which is a bit weird.

**Reply by ghost** [reply]
> Yeah, my original sentence specified 'using type number' and you said it wasn't true. I understand that you don't use it and i get that it offers a poor user experience, but that's what a browser considers as a number.
> 
> By the way, the problem is that `.number` casts to... Well... a js number and in period is the only allowed decimal separator.
> 
> The project is open source so you can try to implement what you described but the property in x-data will end up using a period so when x-model updates, itwill change what the user entered which is a bit weird.

Actually, now that I've changed my browser's language to Hungarian it did accept the comma as a separator for the link you provided, only tried it with Safari so far, try it out if you can. 

I agree it would behave weirdly but to common users who are (and let's be honest there are a lot) not that competent with computers will be confused if they write what they'd consider as a decimal number and the input would show them an integer instead.

**Reply by ghost** [reply]
Okay so I played around a bit and with the browser's language set as Hungarian, the browser correctly handles the value 12,3 as a number. If you check the inputs HTMLInputElement property the "value" attribute holds the input as a string 12,3 while it has the "correct" number 12.3 stored in the "valueAsNumber" attribute as described in: 
https://developer.mozilla.org/en-US/docs/Web/API/HTMLInputElement

I'll close this conv and open up a feature idea.

**Reply by SimoTod** [reply]
> Actually, now that I've changed my browser's language to Hungarian it did accept the comma as a separator for the link you provided, only tried it with Safari so far, try it out if you can.

I can't try safari on mobile because I'm on android but Chrome mobile doesn't seem to do it for me. you are not using type number anyway, it was just to say that browser may or may not support it in certain cases.

If you implement that support, you can maybe use toLocalString to write the number back to x-model using the expected separator.
